File: src/main/java/ru/hollowhorizon/hollowengine/client/screen/widget/FadeInLabelWidget.java
package ru.hollowhorizon.hollowengine.client.screen.widget;

import com.mojang.blaze3d.matrix.MatrixStack;
import net.minecraft.util.text.ITextComponent;
import net.minecraft.util.text.StringTextComponent;
import ru.hollowhorizon.hc.client.screens.widget.HollowWidget;
import ru.hollowhorizon.hc.client.utils.ScissorUtil;

public class FadeInLabelWidget extends HollowWidget {
    public static final int DELAY = 20;
    private int ticker = 0;
    private ITextComponent text = new StringTextComponent("");
    private Runnable onFadeInComplete;
    private boolean paused = true;
    private int lastValue = 0;

    public FadeInLabelWidget(int x, int y, int width, int height) {
        super(x, y, width, height, new StringTextComponent(""));
    }

    public void setText(ITextComponent text) {
        this.text = text;
        this.ticker = 0;
        paused = true;
    }

    public void resetTicker() {
        this.ticker = 0;
        paused = false;
    }

    public int getTicker() {
        return ticker;
    }

    public void onFadeInComplete(Runnable onFadeInComplete) {
        this.onFadeInComplete = onFadeInComplete;
    }

    public void setText(String text) {
        setText(new StringTextComponent(text));
    }

    @Override
    public void renderButton(MatrixStack stack, int mouseX, int mouseY, float ticks) {
        super.renderButton(stack, mouseX, mouseY, ticks);

        int current = (int) (this.width * (this.ticker / (float) DELAY));

        ScissorUtil.start(this.x, this.y, (int) (lastValue + (current - lastValue) * ticks), this.height);

        lastValue = current;

        stack.pushPose();
        stack.translate(0, 0, 500f);
        this.getFont().drawShadow(stack, this.text, this.x, this.y + this.height / 2F - this.getFont().lineHeight / 2F, 0xFFFFFF);
        stack.popPose();
        ScissorUtil.stop();
    }

    @Override
    public void tick() {
        if(paused) return;

        if (this.ticker < DELAY) {
            this.ticker++;
        } else {
            if (this.onFadeInComplete != null) {
                this.onFadeInComplete.run();
            }
            this.paused = true;
        }
    }

    public void setTicker(int ticker) {
        this.ticker = ticker;
    }

    @Override
    public boolean mouseClicked(double mouseX, double mouseY, int button) {
        return false;
    }

    public boolean isComplete() {
        return getFont().width(this.text) <= this.width * this.ticker / DELAY;
    }

    public void complete() {
        this.paused = true;
        this.ticker = DELAY;
    }

    public String getText() {
        return this.text.getString();
    }
}

File: src/main/java/ru/hollowhorizon/hollowengine/client/screen/widget/dialogue/FadeInLabelWidget.kt
package ru.hollowhorizon.hollowengine.client.screen.widget.dialogue

import com.mojang.blaze3d.vertex.PoseStack
import net.minecraft.network.chat.Component
import ru.hollowhorizon.hc.client.screens.widget.HollowWidget
import ru.hollowhorizon.hc.client.utils.GuiAnimator
import ru.hollowhorizon.hc.client.utils.ScissorUtil
import ru.hollowhorizon.hc.client.utils.math.Interpolation
import ru.hollowhorizon.hc.client.utils.mcText


class FadeInLabelWidget(x: Int, y: Int, width: Int, height: Int) : HollowWidget(x, y, width, height, "".mcText) {
    private var text = "".mcText
    private var onFadeInComplete: Runnable? = null
    private var animator: GuiAnimator? = null
    private var isUpdated = false
    override fun init() {
        super.init()
        animator = GuiAnimator.Single(0, width, 1f, Interpolation.EXPO_OUT.function)
    }

    fun setText(text: Component) {
        this.text = text
    }

    fun setText(text: String) {
        setText(text.mcText)
    }

    fun reset() {
        animator?.reset()
    }

    fun onFadeInComplete(onFadeInComplete: Runnable) {
        this.onFadeInComplete = onFadeInComplete
    }

    override fun renderButton(stack: PoseStack, mouseX: Int, mouseY: Int, ticks: Float) {
        super.renderButton(stack, mouseX, mouseY, ticks)
        ScissorUtil.push(x, y, animator!!.value, height)
        stack.pushPose()
        stack.translate(0.0, 0.0, 500.0)
        font.drawShadow(stack, text, x.toFloat(), y + height / 2f - font.lineHeight / 2f, 0xFFFFFF)
        stack.popPose()
        ScissorUtil.pop()
    }

    override fun tick() {
        if (animator?.isFinished() == true && !isUpdated) {
            onFadeInComplete?.run()
            isUpdated = true
        }
    }

    override fun mouseClicked(mouseX: Double, mouseY: Double, button: Int): Boolean {
        return false
    }

    val isComplete: Boolean
        get() = animator!!.isFinished()

    fun complete() {
        animator!!.setTime(1.0f)
    }

    fun getText(): String {
        return text.string
    }
}
